dcicutils/scripts/publish_to_pypi.py

```diff
@@ -0,0 +1,302 @@
+# item #4 above specifically addresses/prevents this. Perhaps better
+# would be if publishing only happened via GitHub actions.
+
+import argparse
+import os
+import requests
+import subprocess
+import toml
+from typing import Tuple, Union
+
+
+DEBUG = False
+PRINT = print
+
+
+def main() -> None:
+
+    def is_github_actions_context():
+        return "GITHUB_ACTIONS" in os.environ
+
+    argp = argparse.ArgumentParser()
+    argp.add_argument("--noconfirm", required=False, dest="noconfirm", action="store_true")
+    argp.add_argument("--debug", required=False, dest="debug", action="store_true")
+    args = argp.parse_args()
+
+    if args.debug:
+        global DEBUG
+        DEBUG = True
+
+    if args.noconfirm and not is_github_actions_context():
+        PRINT("The --noconfirm flag is only allowed within GitHub actions!")
+        exit_with_no_action()
+
+    if not verify_git_repo():
+        exit_with_no_action()
+
+    if not verify_unstaged_changes():
+        exit_with_no_action()
+
+    if not verify_uncommitted_changes():
+        exit_with_no_action()
+
+    if not verify_unpushed_changes():
+        exit_with_no_action()
+
+    if not verify_tagged():
+        exit_with_no_action()
+
+    if not verify_untracked_files():
+        exit_with_no_action()
+
+    package_name = get_package_name()
+    package_version = get_package_version()
+
+    if not verify_not_already_published(package_name, package_version):
+        exit_with_no_action()
+
+    if not args.noconfirm:
+        if not answered_yes_to_confirmation(f"Do you want to publish {package_name} {package_version} to PyPi?"):
+            exit_with_no_action()
+
+    PRINT(f"Publishing {package_name} {package_version} to PyPi ...")
+
+    if not publish_package():
+        exit_with_no_action()
+
+    PRINT(f"Publishing {package_name} {package_version} to PyPi complete.")
+
+
+def publish_package(pypi_username: str = None, pypi_password: str = None) -> bool:
+    if not pypi_username:
+        pypi_username = os.environ.get("PYPI_USER")
+    if not pypi_password:
+        pypi_password = os.environ.get("PYPI_PASSWORD")
+    if not pypi_username or not pypi_password:
+        PRINT(f"No PyPi credentials. You must have PYPI_USER and PYPI_PASSWORD environment variables set.")
+        return False
+    poetry_publish_command = [
+        "poetry", "publish",
+        "--no-interaction", "--build",
+        f"--username={pypi_username}", f"--password={pypi_password}"
+    ]
+    poetry_publish_results, status_code = execute_command(poetry_publish_command)
+    PRINT("\n".join(poetry_publish_results))
+    if status_code != 0:
+        PRINT(f"Publish to PyPi failed!")
+        return False
+    return True
+
+
+def verify_git_repo() -> bool:
+    """
+    If this (the current directory) looks like a git repo then return True,
+    otherwise prints a warning and returns False.
+    """
+    _, status = execute_command("git rev-parse --is-inside-work-tree")
```

Comment:
I won't harp on this too much but there is a GitPython library that may allow you to these calls internally within Python (instead of subprocess which I assume you're doing internally).

Reply:
Hm, yeah, I sorta thought of looking into something like that but figured, at first anyways, to keep it as much like the existing bash script as possible, but maybe this would be better, may look into it, thanks.